README.md
# SMARTER-backend

SMARTER Backend API

[![Build Status](https://www.travis-ci.com/cnr-ibba/SMARTER-backend.svg?branch=master)](https://www.travis-ci.com/cnr-ibba/SMARTER-backend)
[![Coverage Status](https://coveralls.io/repos/github/cnr-ibba/SMARTER-backend/badge.svg?branch=master)](https://coveralls.io/github/cnr-ibba/SMARTER-backend?branch=master)
[![Documentation Status](https://readthedocs.org/projects/smarter-backend/badge/?version=latest)](https://smarter-backend.readthedocs.io/en/latest/)

## Setting the `.env` file

`docker-compose` can read variables from a `.env` placed in the working directory.
Here we will define all variables useful for our containers, like database password.
Edit a new `.env` file in working directory and set passwords for such environment
variables:

```text
MONGODB_ROOT_USER=<root user>
MONGODB_ROOT_PASS=<root pass>
MONGODB_SMARTER_USER=<smarter user>
MONGODB_SMARTER_PASS=<smarter pass>
JWT_SECRET_KEY=<secret key>
```

> *TODO*: manage sensitive data using secret in docker-compose, as described
[here](https://docs.docker.com/engine/swarm/secrets/#use-secrets-in-compose) and
[here](https://docs.docker.com/compose/compose-file/#secrets)

### Genereate the JWT secret key

[Here](https://stackoverflow.com/a/23728630/4385116) you can find hints on
how to define a secret key with python:

```python
import string
import secrets
print(''.join(secrets.choice(string.ascii_letters + string.digits) for _ in range(50)))
```

## Setting the proper permissions

Set `mongodb-home` folder permissions with:

```bash
chmod 777 mongodb-home/
chmod o+t mongodb-home/
```

Fix `flask-data` permissions:

```bash
docker-compose run --no-deps --rm uwsgi sh -c 'chgrp -R www-data .'
cd flask-data/
find . -type f -iname "*.py" -exec chmod g-w {} \;
```

## Add a smarter user

Add a smarter user by calling a *flask script*:

```bash
docker-compose run --rm uwsgi flask users create smarter
```

## Connect to mongodb

```bash
docker-compose run --rm --user mongodb mongo sh -c 'mongo --host mongo --username="${MONGO_INITDB_ROOT_USERNAME}" --password="${MONGO_INITDB_ROOT_PASSWORD}"'
```

## Import data into database

Execute a `mongodump` of an instance of the
[SMARTER-database](https://github.com/cnr-ibba/SMARTER-database) project, for
example:

```bash
docker-compose run --rm --user mongodb mongo sh -c 'DATE=$(date +%Y-%m-%d); mongodump --host mongo --username="${MONGO_INITDB_ROOT_USERNAME}" --password="${MONGO_INITDB_ROOT_PASSWORD}" --authenticationDatabase admin --db=smarter --gzip --archive=/home/mongodb/${DATE}\_smarter.archive.gz'
```

Then copy (or move) the dump file into `mongodb-home` folder of this project. You
can restore the database using `mongorestore`, for example:

```bash
docker-compose run --rm --user mongodb mongo sh -c 'mongorestore --host mongo --username="${MONGO_INITDB_ROOT_USERNAME}" --password="${MONGO_INITDB_ROOT_PASSWORD}" --authenticationDatabase admin --db=smarter --drop --preserveUUID --gzip --archive=/home/mongodb/2021-06-18_smarter.archive.gz'
```

## Monitoring UWSGI processes

Enter inside uwsgi container (with `docker-compose exec`), then monitor uwsgi with
[uwsgitop](https://github.com/xrmx/uwsgitop):

```bash
docker-compose exec uwsgi bash
uwsgitop /tmp/smarter-stats.sock
```

Type `q` to exit from monitoring process

## Test application

Some useful commands to test the application:

```bash
# open the flask shell
docker-compose run --rm uwsgi flask shell

# test and exit when a issue is found. Call first the failed test on successive calls
docker-compose run --rm uwsgi pytest --verbosity=2 --exitfirst --failed-first --showlocals

# test with coverage
docker-compose run --rm uwsgi coverage run --source='.' -m py.test

# generate coverage report locally
docker-compose run --rm uwsgi coverage html

# check code with flake8
docker-compose run --rm uwsgi flake8

# test like CI
docker-compose run --no-deps --rm uwsgi sh -c 'coverage run --source='.' -m py.test && flake8'
```


HISTORY.md

History
=======

TODO
----

* Configure logging [#7](https://github.com/cnr-ibba/SMARTER-backend/issues/7)
* Try to compress responses [#34](https://github.com/cnr-ibba/SMARTER-backend/issues/34)
* Support searching with patterns on original and smarter id [#19](https://github.com/cnr-ibba/SMARTER-backend/issues/19)
* Support `text/csv` header request [#16](https://github.com/cnr-ibba/SMARTER-backend/issues/16)
* Substitute flask-restful components with better alternatives [#10](https://github.com/cnr-ibba/SMARTER-backend/issues/10)

0.2.2 (2023-06-30)

* Search samples by *aliases* [#38](https://github.com/cnr-ibba/SMARTER-backend/issues/38)
* Update models top latest SMARTER database
* Update docker image

0.2.1 (2022-05-11)
------------------

* Fix `dataset` parameter in swagger Samples endpoint
* Add country endpoint [#33](https://github.com/cnr-ibba/SMARTER-backend/issues/33)
* Support multiple parameters in Breeds endpoint [#29](https://github.com/cnr-ibba/SMARTER-backend/issues/29)
* Support parameters into GeoJSON endpoint [#32](https://github.com/cnr-ibba/SMARTER-backend/issues/32)
* Return error when searching variant with a wrong region [#30](https://github.com/cnr-ibba/SMARTER-backend/issues/30)

0.2.0 (2021-12-20)
------------------

* support for GIS search in sample endpoints [#13](https://github.com/cnr-ibba/SMARTER-backend/issues/13)
  * document *GeoJSON* endpoint and GIS parameters using swagger
  * returning all samples as *GeoJSON* FeatureCollection
  * select samples within a sphere
  * select samples inside a polygon
  * model locations with `MultiPointField`
* configure MongoDB memory usage [#12](https://github.com/cnr-ibba/SMARTER-backend/issues/12)
* move base location from `/api/` to `/smarter-api/`
* document API with `flasgger` [#6](https://github.com/cnr-ibba/SMARTER-backend/issues/6)
* add endpoint for supported assemblies variants [#11](https://github.com/cnr-ibba/SMARTER-backend/issues/11)
* add info endpoint
* query with multiple parameters [#23](https://github.com/cnr-ibba/SMARTER-backend/issues/23)
* return 4xx error when querying with wrong parameters [#22](https://github.com/cnr-ibba/SMARTER-backend/issues/22)
* Support `type` argument when searching samples
* Track version with `bump2version`

0.1.0 (2021-11-11)
------------------

* First release of SMARTER-backend
* Support user authentication through JWT
* Start documentation with readthedocs
* Add endpoints for the principal features
  * breeds
  * supported chips
  * datasets
  * samples
  * variants


flask-data/smarter/app.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Fri May 21 17:50:23 2021

@author: Paolo Cozzi <paolo.cozzi@ibba.cnr.it>
"""

import os

from bson import ObjectId
from logging.config import dictConfig

from flask import Flask, redirect, url_for
from flask_restful import Api
from flask_bcrypt import Bcrypt
from flask_jwt_extended import JWTManager
from flask_mongoengine.json import MongoEngineJSONEncoder
from flask_cors import CORS
from flasgger import Swagger

from database.db import initialize_db, DB_ALIAS
from resources.errors import errors
from resources.routes import initialize_routes
from commands import usersbp

__version__ = "0.2.2"

# https://flask.palletsprojects.com/en/2.0.x/logging/#basic-configuration
dictConfig({
    'version': 1,
    'formatters': {'default': {
        'format': '[%(asctime)s] %(levelname)s in %(module)s: %(message)s',
    }},
    'handlers': {'wsgi': {
        'class': 'logging.StreamHandler',
        'stream': 'ext://sys.stdout',
        'formatter': 'default'
    }},
    'root': {
        'level': 'INFO',
        'handlers': ['wsgi']
    }
})


class CustomJSONEncoder(MongoEngineJSONEncoder):
    def default(self, obj):
        if isinstance(obj, ObjectId):
            return {
                "$oid": str(obj)
            }
        return MongoEngineJSONEncoder.default(self, obj)


# https://stackoverflow.com/a/56474420/4385116
def create_app(config={}):
    """This function create Flask app. Is required by wsgi because it need to
    be called after service is started and forked, not when importing the
    module during initialization. To start the flask app, first import
    the module and then create all the stuff by invoking this function
    You need call the run method on the returned values to start acception
    requests

    Args:
        config (dict): pass parameters to this app (not yet defined)

    Returns:
        Flask: a flask initialized application
    """

    app = Flask(__name__)
    CORS(app)
    api = Api(app, errors=errors)
    Bcrypt(app)
    JWTManager(app)

    # deal with ObjectId in json responses
    app.json_encoder = CustomJSONEncoder

    # workaround to make flasgger deal with jwt-token headers
    app.config["JWT_AUTH_URL_RULE"] = True

    # Swagger stuff
    swagger_template = {
        "swagger": "2.0",
        "info": {
            "title": "SMARTER-backend API",
            "description": "REST API for SMARTER data",
            "termsOfService": None,
            "version": __version__
        },
        "basePath": "/smarter-api/",  # base bash for blueprint registration
    }

    swagger_config = {
        "headers": [],
        "specs": [
            {
                "endpoint": '/smarter-api/apispec_1',
                "route": '/smarter-api/apispec_1.json',
                "rule_filter": lambda rule: True,  # all in
                "model_filter": lambda tag: True,  # all in
            }
        ],
        "static_url_path": "/smarter-api/flasgger_static",
        # "static_folder": "static",  # must be set by user
        "swagger_ui": True,
        "specs_route": "/smarter-api/docs/"
    }

    Swagger(app, template=swagger_template, config=swagger_config)

    app.logger.debug("App initialized")

    # http://docs.mongoengine.org/projects/flask-mongoengine/en/latest/#configuration
    app.config['MONGODB_SETTINGS'] = {
        'host': 'mongodb://mongo/smarter',
        'username': os.getenv("MONGODB_SMARTER_USER"),
        'password': os.getenv("MONGODB_SMARTER_PASS"),
        'authentication_source': 'admin',
        'alias': DB_ALIAS,
        # NOTE: This fixes "UserWarning: MongoClient opened before fork."
        # I'm not aware of side effects yet. Default value is/was "True"
        'connect': False
    }

    # override configuration with custom values
    if 'host' in config:
        app.logger.error(f"Setting custom host: {config['host']}")
        app.config['MONGODB_SETTINGS']['host'] = config['host']

    # https://flask-jwt-extended.readthedocs.io/en/stable/basic_usage/
    app.config["JWT_SECRET_KEY"] = os.getenv('JWT_SECRET_KEY')

    # connect to database
    initialize_db(app)

    app.logger.debug("Database initialized")

    # you MUST register the blueprint
    app.register_blueprint(usersbp)

    # add resources
    initialize_routes(api)

    app.logger.debug("Routes initialized")

    # add a redirect for the index page
    @app.route('/smarter-api/')
    def index():
        return redirect(url_for('flasgger.apidocs'))

    return app


if __name__ == '__main__':
    app = create_app()
    app.run()
